src/macaron/json_tools.py

```diff
@@ -34,12 +33,22 @@ def json_extract(entry: JsonType, keys: list[str], type_: type[T]) -> T | None:
     target = entry
 
     for index, key in enumerate(keys):
-        if not isinstance(target, dict):
-            logger.debug("Expect the value .%s to be a dict.", ".".join(keys[:index]))
-            return None
-        if key not in target:
-            logger.debug("JSON key '%s' not found in .%s", key, ".".join(keys[:index]))
+        if isinstance(target, dict):
+            if key not in target:
+                logger.debug("JSON key '%s' not found in .%s", key, ".".join(keys[:index]))
+                return None
+        elif isinstance(target, list):
+            if isinstance(key, str):
+                logger.debug("JSON key '%s' of type 'str' cannot extract from list.", key)
+                return None
+
+            if key < 0 or key >= len(target):
+                logger.debug("JSON key '%s' is outside of list bounds %s.", key, len(target))
```

Comment:
nit: I think it should be called `JSON list index`  in the debug message :thinking: